src/vos/tests/vts_aggregate.c

```diff
@@ -0,0 +1,878 @@
+/**
+ * (C) Copyright 2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * This file is part of vos/tests/
+ *
+ * vos/tests/vts_aggregate.c
+ */
+#define D_LOGFAC	DD_FAC(tests)
+
+#include "vts_io.h"
+#include <vos_internal.h>	/* for VOS_BLK_SZ */
+
+#define VERBOSE_MSG(...)			\
+{						\
+	if (false)				\
+		print_message(__VA_ARGS__);	\
+}
+
+static void
+update_value(struct io_test_args *arg, daos_unit_oid_t oid, daos_epoch_t epoch,
+	     char *dkey, char *akey, daos_iod_type_t type, daos_size_t iod_size,
+	     daos_recx_t *recx, char *buf)
+{
+	daos_iod_t	iod = { 0 };
+	daos_sg_list_t	sgl = { 0 };
+	daos_key_t	dkey_iov, akey_iov;
+	daos_size_t	buf_len;
+	int		rc;
+
+	assert_true(dkey != NULL && akey != NULL);
+	assert_true(strlen(dkey) && strlen(akey));
+	assert_true(!(arg->ta_flags & TF_ZERO_COPY));
+
+	arg->oid = oid;
+	daos_iov_set(&dkey_iov, dkey, strlen(dkey));
+	daos_iov_set(&akey_iov, akey, strlen(akey));
+
+	rc = daos_sgl_init(&sgl, 1);
+	assert_int_equal(rc, 0);
+
+	if (type == DAOS_IOD_SINGLE)
+		buf_len = iod_size;
+	else
+		buf_len = recx->rx_nr * iod_size;
```

Comment:
Mostly for my own understanding of how things work, but ... if `type == DAOS_IOD_SINGLE` then `recx->rx_nr == 1`, correct? In that case the conditional really isn't needed, but it is probably more explicit, so I'm not necessarily suggesting changing it.

Reply:
I checked DAOS API & server I/O code, looks there is such requirement that 'recx' must be provided for IOD_SINGLE, so in this test program, 'recx' will be NULL for IOD_SINGLE.